sklearn/utils/estimator_checks.py

```diff
@@ -1196,6 +1197,61 @@ def check_array_api_input_and_values(
     )
 
 
+def check_estimator_sparse_tag(name, estimator_orig):
+    """Check that estimator tag related with accepting sparse data is properly set."""
+    if SPARSE_ARRAY_PRESENT:
+        sparse_container = sparse.csr_array
+    else:
+        sparse_container = sparse.csr_matrix
+    estimator = clone(estimator_orig)
+
+    rng = np.random.RandomState(0)
+    n_samples = 15 if name == "SpectralCoclustering" else 40
+    X = rng.uniform(size=(n_samples, 3))
+    X[X < 0.6] = 0
+    y = rng.randint(0, 3, size=n_samples)
+    X = _enforce_estimator_tags_X(estimator, X)
+    y = _enforce_estimator_tags_y(estimator, y)
+    X = sparse_container(X)
+
+    tags = get_tags(estimator)
+    if tags.input_tags.sparse:
+        try:
+            estimator.fit(X, y)  # should pass
+        except Exception as e:
+            raise AssertionError(
+                f"Estimator {name} raised an exception: {e}. The tag "
+                "self.input_tags.sparse might not be consistent with the "
+                "estimator's ability to handle sparse data (i.e. controlled by the "
+                "parameter `accept_sparse` in `validate_data` or `check_array` "
+                f"functions). Got input_tags.sparse={tags.input_tags.sparse}."
+            )
```

Comment:
Should I maybe `raise AssertionError(err_msg) from e` as we did for the `tags.input_tags.sparse=False` case ?

Reply:
Yes I like this suggestion to have a proper traceback. This is a developer traceback so we can be verbose at this stage.

Reply:
done in 425b473